File: C++/Problems/Problems/problem_3325.cpp
#include "problem_3325.h"

bool problem_3325::test() {
    std::string s = "abacb";
    int k = 2;

    int expected = 4;

    int result = numberOfSubstrings(s, k);

    std::cout << result << std::endl;

    return expected == result;
}

int problem_3325::numberOfSubstrings(std::string s, int k) {
    int n = s.size();
    int i = -1;
    int j = -1;
    int result = 0;
    int maxLetter = 0;
    std::vector counts(26, 0);
    while (i < n - 1)
    {
        while (i < n - 1 && counts[maxLetter] < k)
        {
            i++;
            counts[s[i] - 'a']++;
            if (counts[s[i] - 'a'] > counts[maxLetter])
                maxLetter = s[i] - 'a';
        }

        while (j < n - 1 && counts[maxLetter] >= k)
        {
            result += (n - i);
            j++;
            counts[s[j] - 'a']--;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem3325.py
from typing import List

from .problem import Problem


class Problem3325(Problem):
    def test(self) -> bool:
        s = "abacb"
        k = 2

        expected = 4

        result = self.numberOfSubstrings(s, k)

        print(result)

        return expected == result

    def numberOfSubstrings(self, s: str, k: int) -> int:
        n = len(s)
        i = -1
        j = -1
        result = 0
        maxLetter = 0
        counts = [0] * 26
        while i < n - 1:
            while i < n - 1 and counts[maxLetter] < k:
                i += 1
                counts[ord(s[i]) - ord('a')] += 1
                if counts[ord(s[i]) - ord('a')] > counts[maxLetter]:
                    maxLetter = ord(s[i]) - ord('a')

            while j < n - 1 and counts[maxLetter] >= k:
                result += (n - i)
                j += 1
                counts[ord(s[j]) - ord('a')] -= 1

        return result